Mathlib/Analysis/ODE/Basic.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2023 Winston Yin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Winston Yin
+-/
+module
+
+public import Mathlib.Analysis.Calculus.Deriv.Basic
+
+/-!
+# Integral curves of vector fields on a Banach space
+
+Let `E` be a Banach space and `v : E → E` be a vector field on `E`. An integral curve of `v` is a
+function `γ : ℝ → E` such that the derivative of `γ` at `t` equals `v t (γ t)`. The integral curve
+may only be defined for all `t` within some subset of `ℝ`.
+
+## Main definitions
+
+Let `v : E → E` be a vector field on `E`, and let `γ : ℝ → E`.
+* `IsIntegralCurve γ v`: `γ t` is tangent to `v (γ t)` for all `t : ℝ`. That is, `γ` is a global
+  integral curve of `v`.
+* `IsIntegralCurveOn γ v s`: `γ t` is tangent to `v (γ t)` for all `t ∈ s`, where `s : Set ℝ`.
+* `IsIntegralCurveAt γ v t₀`: `γ t` is tangent to `v (γ t)` for all `t` in some open interval
+  around `t₀`. That is, `γ` is a local integral curve of `v`.
+
+For `IsIntegralCurveOn γ v s` and `IsIntegralCurveAt γ v t₀`, even though `γ` is defined for all
+time, its value outside of the set `s` or a small interval around `t₀` is irrelevant and considered
+junk.
+
+## TODO
+
+* Implement `IsIntegralCurveWithinAt`.
+
+## Tags
+
+integral curve, vector field
+-/
+
+@[expose] public section
+
+open scoped Topology
+
+open Set
+
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E]
+
+/-- `IsIntegralCurveOn γ v s` means `γ t` is tangent to `v (γ t)` within `s` for all `t ∈ s`. The
+value of `γ` outside of `s` is irrelevant and considered junk. -/
+def IsIntegralCurveOn (γ : ℝ → E) (v : E → E) (s : Set ℝ) : Prop :=
+  ∀ t ∈ s, HasDerivWithinAt γ (v (γ t)) s t
+
+/-- `IsIntegralCurveAt γ v t₀` means `γ : ℝ → E` is a local integral curve of `v` in a neighbourhood
+containing `t₀`. The value of `γ` outside of this neighbourhood is irrelevant and considered
+junk. -/
+def IsIntegralCurveAt (γ : ℝ → E) (v : E → E) (t₀ : ℝ) : Prop :=
+  ∀ᶠ t in 𝓝 t₀, HasDerivAt γ (v (γ t)) t
+
+/-- `IsIntegralCurve γ v` means `γ : ℝ → E` is a global integral curve of `v`. That is, `γ t` is
+tangent to `v (γ t)` for all `t : ℝ`. -/
+def IsIntegralCurve (γ : ℝ → E) (v : E → E) : Prop :=
+  ∀ t : ℝ, HasDerivAt γ (v (γ t)) t
+
+variable {γ γ' : ℝ → E} {v : E → E} {s s' : Set ℝ} {t₀ : ℝ}
+
+lemma IsIntegralCurve.isIntegralCurveOn (h : IsIntegralCurve γ v) (s : Set ℝ) :
+    IsIntegralCurveOn γ v s := fun t _ ↦ (h t).hasDerivWithinAt
+
+lemma isIntegralCurve_iff_isIntegralCurveOn :
+    IsIntegralCurve γ v ↔ IsIntegralCurveOn γ v univ :=
+  ⟨fun h ↦ h.isIntegralCurveOn _, fun h t ↦ (h t (mem_univ _)).hasDerivAt Filter.univ_mem⟩
+
+lemma isIntegralCurveAt_iff :
```

Comment:
This lemma name is not very good, as there could be many equivalent characterizations of integral curves. `isIntegralCurveAt_iff_exists_mem_nhds`?